drivers/storage/portworx/status.go

```diff
@@ -468,30 +467,27 @@ func mapNodeStatus(status api.Status) corev1.NodeConditionStatus {
 }
 
 func getStorageNodePhase(status *corev1.NodeStatus) string {
-	latestTime := metav1.NewTime(time.Time{})
-	var latestCondition *corev1.NodeCondition
+	var nodeInitCondition *corev1.NodeCondition
+	var nodeStateCondition *corev1.NodeCondition
 
 	for _, condition := range status.Conditions {
-		// Find the latest condition. If it is InitCondition, and has
-		// the same timestamp as the latest one then don't make it latest
-		if latestTime.Before(&condition.LastTransitionTime) ||
-			latestTime.IsZero() ||
-			(latestTime.Equal(&condition.LastTransitionTime) &&
-				condition.Type != corev1.NodeInitCondition) {
-			latestCondition = condition.DeepCopy()
-			latestTime = condition.LastTransitionTime
+		if condition.Type == corev1.NodeInitCondition {
+			nodeInitCondition = condition.DeepCopy()
+		} else if condition.Type == corev1.NodeStateCondition {
+			nodeStateCondition = condition.DeepCopy()
 		}
 	}
 
-	// If no condition or status found return Initializing phase.
-	// Also if the InitCondition is the latest condition and it has succeeded,
-	// then keep the node phase as Initializing
-	if latestCondition == nil || latestCondition.Status == "" ||
-		(latestCondition.Type == corev1.NodeInitCondition &&
-			latestCondition.Status == corev1.NodeSucceededStatus) {
+	if nodeInitCondition == nil || nodeInitCondition.Status == corev1.NodeSucceededStatus {
+		if nodeStateCondition != nil && nodeStateCondition.Status != "" {
+			return string(nodeStateCondition.Status)
+		}
 		return string(corev1.NodeInitStatus)
+	} else if nodeStateCondition == nil ||
+		nodeStateCondition.LastTransitionTime.Before(&nodeInitCondition.LastTransitionTime) {
```

Comment:
Is it possible the evaluation of the expression be true? 
`nodeStateCondition.LastTransitionTime.Before(&nodeInitCondition.LastTransitionTime)`

Reply:
This was the main issue in the previous code. 
The NodeInit condition is added by portworx, while the NodeState is added by operator. There was a race when portworx restarted quickly and updated the NodeInit condition, but as operator thought px was Online it never updated the NodeState condition. So the NodeInit condition had a later timestamp than the NodeState condition.

Reply:
I see. It's due to the px restart. It's confusing when I first see the equation. I would suggest you put a comment to explain why the NodeInit condition had a later timestamp than the NodeState condition.

Again, the state machine can avoid checking the timestamp. Both PX and Operator can change state, and the operator just needs to do a periodical check in reconciling loop to take action based on the state and conditions.  
Our PSO cockroachDB operator also has similar problems, this is why I wrote a PoC operator.

Reply:
Sounds good. I will add a comment and we can integrate the state machine for conditions later.